src/shogun/preprocessor/RandomFourierGaussPreproc.cpp

```diff
@@ -8,393 +8,131 @@
-	randomcoeff_additive=NULL;
-
-	cur_dim_feature_space=dim_feature_space;
-	cur_dim_input_space = dim_input_space;
-	cur_kernelwidth=kernelwidth;
-
-	if( (dim_feature_space>0) && (dim_input_space>0) )
-	{
-	randomcoeff_additive=SG_MALLOC(float64_t, cur_dim_feature_space);
-	randomcoeff_multiplicative=SG_MALLOC(float64_t, cur_dim_feature_space*cur_dim_input_space);
-
-	std::copy(randomcoeff_additive2, randomcoeff_additive2
-			+ dim_feature_space, randomcoeff_additive);
-	std::copy(randomcoeff_multiplicative2, randomcoeff_multiplicative2
-			+ cur_dim_feature_space*cur_dim_input_space, randomcoeff_multiplicative);
-	}
+void RandomFourierGaussPreproc::init_basis(int32_t dim_input_space)
+{
+	io::info("generating random coefficients {}x{}",m_dims,dim_input_space);
+	m_basis = SGMatrix<float64_t>(m_dims,dim_input_space);
+	m_offset = SGVector<float64_t>(m_dims);
 
+	sample_spectral_density(dim_input_space);
 }
 
 void RandomFourierGaussPreproc::fit(std::shared_ptr<Features> f)
 {
-	if (dim_feature_space <= 0) {
-		throw ShogunException(
-				"CRandomFourierGaussPreproc::init (Features *f): dim_feature_space<=0 is not allowed, use void set_dim_feature_space(const int32 dim) before!\n");
-	}
+	require(m_dims > 0, "Feature space dimension must be greater than zero");
+	require(m_kernelwidth > 0, "Kernelwidth must be greater than zero");
 
-	io::info("calling CRandomFourierGaussPreproc::init(...)");
-	int32_t num_features =
-	    f->as<DenseFeatures<float64_t>>()->get_num_features();
+	auto num_features = f->as<DenseFeatures<float64_t>>()->get_num_features();
 
-	if (!test_rfinited()) {
-		dim_input_space = num_features;
-		init_randomcoefficients();
-		ASSERT( test_rfinited())
-	} else {
-		dim_input_space = num_features;
-		// does not reinit if dimension is the same to avoid overriding a previous call of set_randomcoefficients(...)
-		init_randomcoefficients();
-	}
+	require(num_features > 0, "Input space dimension must be greater than zero");
+
+	init_basis(num_features);
+	m_fitted = true;
 }
 
 SGVector<float64_t> RandomFourierGaussPreproc::apply_to_feature_vector(SGVector<float64_t> vector)
 {
-	if (!test_rfinited()) {
-		throw ShogunException(
-				"float64_t * CRandomFourierGaussPreproc::apply_to_feature_vector(...): test_rfinited()==false: you need to call before CRandomFourierGaussPreproc::init (Features *f) OR	1. set_dim_feature_space(const int32 dim), 2. set_dim_input_space(const int32 dim), 3. init_randomcoefficients() or set_randomcoefficients(...) \n");
-	}
+	SGVector<float64_t> res(m_dims);
+	SGVector<float64_t> offset_vector(m_dims);
 
-	float64_t val = std::sqrt(2.0 / cur_dim_feature_space);
-	SGVector<float64_t> res(cur_dim_feature_space);
-
-	for (int32_t od = 0; od < cur_dim_feature_space; ++od) {
-		SGVector<float64_t> wrapper(randomcoeff_multiplicative+od*cur_dim_input_space, cur_dim_input_space, false);
-		res[od] = val * cos(randomcoeff_additive[od] + linalg::dot(vector, wrapper));
-	}
+	assert_fitted();
```

Comment:
does the base class take care of flipping the member (is it defined in the base class?)

Reply:
Here's what assert_fitted expands to
```c++
		require<NotFittedException>(
		    m_fitted, "Transformer has not been fitted.");
```

Reply:
In the Transformer class

Reply:
does the base class take care of flipping the member ?

Reply:
Not sure what you mean "flipping the member"

Reply:
changing the value of `m_fitted` when calling `fit`

Reply:
Yes, m_fitted defaults to false and turns true when fit is completed

Reply:
so then you don't need to do that here in this class right? You also don't need to call `assert_fitted` because that is all done in the base class

Reply:
Ah no, Transformer class leaves it's fit method blank.

Reply:
> does the base class take care of flipping the member (is it defined in the base class?)

TLDR; No.

Reply:
ok then leave the assert and flip the flag in here

Reply:
Bad idea I think, we won't be able to call ```apply_``` more than one matrix/vector this way. Better let m_fitted default to false, turning true only when init success. Thoughts?

Reply:
sorry with "in here" I dont mean this method, but this class (i.e. in train)